File: algorithms/[1584]连接所有点的最小费用/solution.ts
// [顶点1索引, 顶点2索引, 权重]
export type edge = [number, number, number];

export class UnionFind {
  parent: number[];
  rank: number[];
  constructor(n: number) {
    this.parent = Array.from({ length: n }, (_, i) => i);
    this.rank = Array.from({ length: n }, () => 1);
  }
  union(i: number, j: number): void {
    const x = this.find(i);
    const y = this.find(j);

    if (this.rank[x] < this.rank[y]) {
      this.parent[x] = y;
    } else {
      this.parent[y] = x;
    }

    if (this.rank[x] === this.rank[y] && x !== y) {
      this.rank[y] += 1;
    }
  }
  find(x: number): number {
    if (this.parent[x] !== x) {
      this.parent[x] = this.find(this.parent[x]);
    }
    return this.parent[x];
  }
}

/**
 * @param {number[][]} points
 * @return {number}
 */
function minCostConnectPoints(points: number[][]): number {
  // 【最小生成树问题】 Kruskal 算法
  // 图解 https://blog.csdn.net/luoshixian099/article/details/51908175
  const n = points.length;
  const uf = new UnionFind(n);
  const edges: edge[] = [];

  for (let i = 0; i < n; i += 1) {
    for (let j = i + 1; j < n; j += 1) {
      const [x1, y1] = points[i];
      const [x2, y2] = points[j];
      edges.push([i, j, Math.abs(x1 - x2) + Math.abs(y1 - y2)]);
    }
  }

  // 所有边按权重（曼哈顿距离）升序排列
  edges.sort((edge1, edge2) => edge1[2] - edge2[2]);

  // 记录构成最小生成树的每条边的权重（曼哈顿距离）
  const res: number[] = [];

  while (res.length < n - 1) {
    const [x, y, w] = edges.shift()!;
    if (uf.find(x) !== uf.find(y)) {
      uf.union(x, y);
      res.push(w);
    }
  }

  return res.reduce((acc, cur) => acc + cur, 0);
};

File: algorithms/[1584]连接所有点的最小费用/solution.py
# topics = ["图", "最小生成树"]

from typing import List, Tuple


class Solution:
    def minCostConnectPoints(self, points: List[List[int]]) -> int:
        """
        【最小生成树】 Prim 算法（贪心）
        【参考】 https://blog.csdn.net/luoshixian099/article/details/51908175
        time O(n^2), space O(n^2), n 为节点数
        """
        # 所有顶点数
        n = len(points)
        # 最小总费用
        min_cost = 0
        # 记录加入最小生成树中的顶点的索引（以第一个点为初始）
        u = [0]
        # 记录所有顶点到最小生成树的最小费用（曼哈顿距离）
        v = [self.get_hamanton_dis(points[0], points[i]) for i in range(n)]

        while len(u) != n:
            # 下一个加入最小生成树的顶点的费用和索引
            next_val, next_idx = self.get_min(v)
            min_cost += next_val
            u.append(next_idx)

            # 更新最小费用数组
            for idx, val in enumerate(v):
                v[idx] = min(val, self.get_hamanton_dis(points[idx], points[next_idx]))

        return min_cost

    def get_hamanton_dis(self, p1: List[int], p2: List[int]) -> int:
        """
        求两点之间的哈曼顿距离
        """
        x1, y1 = p1
        x2, y2 = p2
        return abs(x1 - x2) + abs(y1 - y2)

    def get_min(self, list: List[int]) -> Tuple[int, int]:
        """
        求数组中的最小非 0 值及索引
        """
        min_val = min([val for val in list if val])
        min_idx = list.index(min_val)
        return (min_val, min_idx)
